Update 2i2c access management (#326)
NASA-Openscapes/earthdata-cloud-cookbook: leading-workshops/add-folks-to-2i2c-github-teams.qmd

---
title: "How to Add Folks to the 2i2c Hub"
---

<!--
Move out of leading-workshops - Create "Administration" chapter
Three patterns: 
- One-offs (Long-Term-Access in NASA-Openscapes)
- Workshops
- Cohorts
-->

We use GitHub Teams to manage access to the 2i2c Hub. There are three different patterns
of access:

1. [Individually-managed long-term access for NASA-Openscapes mentors and others](#individual-long-term-access)
2. Access for participants in a NASA-Openscapes champions cohort
3. Access for participants in a workshop

Access for for cohorts and workshop participants currently uses the [same 
workflow](#adding-workshop-participants-or-champions-cohorts-as-a-batch).

We are also working on a method to give just-in-time access for specific workshops

> without using GitHub teams or other managed access. Details of this process will 

Reviewer: ```suggestion
without using GitHub Teams or other managed access. Details of this process will be
```

### 2. Google Form response setup

Set the form responses in [Openscapes_JupyterHubAccess](https://docs.google.com/forms/d/1P_EudQKv3VzoceNSZLv1a3hHM59bZNIvCnwDEeKKvY4/edit#response=ACYDBNiewOAkYG0UxG2XcAmnfnWclek1AAtHeze3mxyaFBF0znCbwqf1TZA6IMHrkGeOXJo) to notify when there are responses: Click on Responses \> click three vertical dots and then check "email notification."

![](images/google-form.png){fig-alt="alt text goes here" fig-align="center" width="391"}

![](images/google-form-notifications.png){fig-alt="alt text goes here" fig-align="center" width="419"}

### 3. Email alert of new signup

When you get an email click '**view response**' (not `view summary`)

![](images/google-form-view-response.png){fig-alt="alt text goes here" fig-align="center" width="436"}

### 4. Copy GitHub username

On the response page, copy the Github Username ![](images/google-form-gh-username.png){fig-alt="alt text goes here" width="406"}

### 5. Add GitHub username to the GitHub Team

::: {.callout-tip}
If you are adding many new mentors to the long-term access team, you can use the 
same steps as [for adding workshop participants](#4-add-users-to-a-team-in-a-batch-using-kyber)
:::

> Go to the NASA Openscapes [LongtermAccess-2i2c Team](https://github.com/orgs/nasa-openscapes/teams/longtermaccess-2i2c) in GitHub \> Click Add member \> Paste GitHub username. Note there are separate [GitHub Teams](https://github.com/orgs/nasa-openscapes-workshops/teams) for Champions Cohorts and individual workshops.

Reviewer: ```suggestion
Go to the NASA Openscapes [LongtermAccess-2i2c Team](https://github.com/orgs/nasa-openscapes/teams/longtermaccess-2i2c) in GitHub \> Click Add member \> Paste GitHub username. Note there are separate [GitHub Teams](https://github.com/orgs/nasa-openscapes-workshops/teams) for Champions Cohorts and workshops.
```

![](images/github-notification-banner-part2.png){fig-align="center" width="85%"}

### 6. Email confirmation

> Go back to the form and grab their email address. Send the following emails to the person.

Reviewer: ```suggestion
Go back to the Google form response and grab their email address. Send the following emails to the person.
```

> **Email for General requests:** 

Reviewer: add new line before "Hi [FIRST NAME]

![](images/github-notification-banner-part2.png){fig-align="center" width="85%"}

### 6. Email confirmation

Go back to the Google form response and grab their email address. Send the following emails to the person.

**Email for General requests:** 

> Subject: NASA Openscapes Access to 2i2c Hub
> 
> Hi \[FIRST NAME\],
>
> I have added you to the NASA Openscapes 2i2c Jupyter Hub. Here is the link to the hub: <https://openscapes.2i2c.cloud/> 
>
> There is a getting started guide in the NASA Earthdata Cloud Cookbook here: <https://nasa-openscapes.github.io/earthdata-cloud-cookbook/>
>

> > We'd love to know about the kind of work you are doing on the hub, and I'll follow up in the next few months. 

Reviewer 2: ```suggestion
> We'd love to know about the kind of work you are doing on the hub. Please add a brief description of your progress as you go at <https://github.com/NASA-Openscapes/2i2cAccessPolicies/discussions/2>. We will follow up in the next few months. 
```

### 1. Create a team in [nasa-openscapes-workshops](https://github.com/orgs/nasa-openscapes-workshops/teams)

There are several teams in this organization; the `AdminTeam` team is for 
members who have permission to create teams and add members to teams.

- If this is for a one-off workshop, name the team `[workshop-name]-[workshop-date]`, with 
workshop date in the format `yyyy-mm-dd`
- If this is for a new champions cohort, name the team `nasa-champions-yyyy`

### 2. Add team name to workshop registry

We keep a registry of all workshops in [this sheet](https://docs.google.com/spreadsheets/d/1Sg4fMDGxMLI0p5cgK5xyamFJpKC6NyNneqXLDReSKcg/edit#gid=695033382). Add the name 
of the GitHub team to `2i2c access type` column in the appropriate row.

### 3. Add teams to 2i2c Hub configuration

Make a pull request to the Openscapes [common.values.yaml](https://github.com/2i2c-org/infrastructure/blob/main/config/clusters/openscapes/common.values.yaml) file in 
the 2i2c infrastructure repository, adding the new team to the `allowed-teams` 

> item in whichever sections are appropriate for your workshop/cohort:

Reviewer 2: ```suggestion
item in whichever sections are appropriate for your workshop/cohort (likely all of them):
```

You also need to add the team to the `allowed_organizations` item under 
`hub > config > GitHubOAuthenticator`

You can see an example pull request [here](https://github.com/2i2c-org/infrastructure/pull/3944).

### 4. Share Google Form with participants needing access for the cohort or workshop

Ask users: Please add your GitHub Username to this [Google form](https://forms.gle/htV2PxJwRaP8jDpC6) 
for NASA Openscapes JupyterHub Access.

### 5. Add users to a Team in a batch using [kyber](https://github.com/Openscapes/kyber)

::: {.callout-tip}
Do this fairly close to the day of the workshop to try to get the greatest 
number particpants to have filled out the form, and reduce the number of manual 
additions on the day of the workshop.
:::

1. Open the [spreadsheet of form responses](https://docs.google.com/spreadsheets/d/1TgokoQa-jYncYSOjeYoWYRuDAbyo4uUR-MSzUotvM00/edit?resourcekey#gid=1416760746), select the GitHub usernames from the rows containing the people you want to add, and copy it to your clipboard.

::: {.callout-tip}
Some workshop instructors will use their own form to record participants,

> rather than the Openscapes form. In that case use those responses instead.

Reviewer 2: ```suggestion
rather than the Openscapes form. In that case use those responses instead. But do paste all those rows into this Main form; it is important that we have a record of everyone using the Hub and a way to contact them via email.
```

:::

2. In R, do the following:

  ```r
  library(kyber)
  library(datapasta)
  

>   # Run datapasta::tribble_paste() or use the RStudio Addin 'Paste as tribble'

Reviewer 2: Note to update this when Andy has time to create a helper function/remove datapasta :)

Author: Note added https://github.com/Openscapes/kyber/issues/88#issuecomment-2135996235

> There is a getting started guide in the NASA Earthdata Cloud Cookbook here: <https://nasa-openscapes.github.io/earthdata-cloud-cookbook/>
>
> We'd love to know about the kind of work you are doing on the hub, and I'll follow up in the next few months. 
>
> Best,

### 7. Register date in Form Response Sheet

> In the [Form Response Sheet](https://docs.google.com/spreadsheets/d/1TgokoQa-jYncYSOjeYoWYRuDAbyo4uUR-MSzUotvM00/edit?resourcekey#gid=1416760746), in Column G add the date they were added to the Hub, or if the request came via 

Reviewer: ```suggestion
In the [Form Response Sheet](https://docs.google.com/spreadsheets/d/1TgokoQa-jYncYSOjeYoWYRuDAbyo4uUR-MSzUotvM00/edit?resourcekey#gid=1416760746), in column `Added to NASA 2i2c hub` add the date they were added to the Hub, or if the request came via 
```

3.  Not seeing that bar? 

    1.  Are you logged into [GitHub.com](https://github.com)?

    2.  Have you shared your GitHub username via [this form](https://forms.gle/XRPsKiqaZC4bkYoi8)?

    3.  You can also check your email that you use for GitHub and look for an invitation from GitHub and NASA-Openscapes

## Adding workshop participants or Champions cohorts as a batch

Participants in workshops run by NASA Openscapes Mentors are given Openscapes 
2i2c JupyterHub access, as are participants in Champions Cohorts.

We use a dedicated GitHub Organization - [nasa-openscapes-workshops](https://github.com/nasa-openscapes-workshops) - to manage access, with [GitHub Teams](https://github.com/orgs/nasa-openscapes-workshops/teams) for workshops and Champions Cohorts.

### 1. Create a team in [nasa-openscapes-workshops](https://github.com/orgs/nasa-openscapes-workshops/teams)

There are several teams in this organization; the `AdminTeam` team is for 
members who have permission to create teams and add members to teams.

- If this is for a one-off workshop, name the team `[workshop-name]-[workshop-date]`, with 
workshop date in the format `yyyy-mm-dd`

> - If this is for a new champions cohort, name the team `nasa-champtions-yyyy`

Reviewer: ```suggestion
- If this is for a new champions cohort, name the team `nasa-champions-yyyy`
```

Reviewer: Notice the Champions team name is `nasa-champions-yyyy` and their repo name would be `yyyy-nasa-champions`. Could they be named the same to reduce confusion?

Author: This makes sense to me!

Author: @jules32 what do you think?

### 2. Add team name to workshop registry

We keep a registry of all workshops in [this sheet](https://docs.google.com/spreadsheets/d/1Sg4fMDGxMLI0p5cgK5xyamFJpKC6NyNneqXLDReSKcg/edit#gid=695033382). Add the name 
of the GitHub team to `2i2c access type` column in the appropriate row.

### 3. Add teams to 2i2c Hub configuration

Make a pull request to the Openscapes [common.values.yaml](https://github.com/2i2c-org/infrastructure/blob/main/config/clusters/openscapes/common.values.yaml) file in 
the 2i2c infrastructure repository, adding the new team to the `allowed-teams` 
item in whichever sections are appropriate for your workshop/cohort:

- `display_name: Python`
- `display_name: R`
- `display_name: Matlab`
- `display_name: "Bring your own image"`

These will determine which images are available to participants when they log in 

> to the hub.

Reviewer: ```suggestion
to the Hub.
```

Team specification is in the form `[github-org]:[team]` (e.g., `nasa-openscapes-workshops:champions-access-2i2c-2024`)

You also need to add the team to the `allowed_organizations` item under 
`hub > config > GitHubOAuthenticator`

You can see an example pull request [here](https://github.com/2i2c-org/infrastructure/pull/3944).

### 4. Share Google Form with participants needing access for the cohort or workshop

Ask users: Please add your GitHub Username to this [Google form](https://forms.gle/htV2PxJwRaP8jDpC6) 
for NASA Openscapes JupyterHub Access.

> ### 5. Add users to a team in a batch using [kyber](https://github.com/Openscapes/kyber)

Reviewer: ```suggestion
### 5. Add users to a Team in a batch using [kyber](https://github.com/Openscapes/kyber)
```

::: {.callout-tip}
Do this fairly close to the day of the workshop to try to get the greatest 
number particpants to have filled out the form, and reduce the number of manual 
additions on the day of the workshop.
:::

> 1. Open the [spreadsheet of form responses](https://docs.google.com/spreadsheets/d/1TgokoQa-jYncYSOjeYoWYRuDAbyo4uUR-MSzUotvM00/edit?resourcekey#gid=1416760746), select the rows containing the people you want to add, and copy it to your clipboard.

Reviewer: ```suggestion
1. Open the [spreadsheet of form responses](https://docs.google.com/spreadsheets/d/1TgokoQa-jYncYSOjeYoWYRuDAbyo4uUR-MSzUotvM00/edit?resourcekey#gid=1416760746), select the GitHub usernames from the rows containing the people you want to add, and copy it to your clipboard.
```

::: {.callout-tip}
Some workshop instructors will use their own form to record participants,
rather than the Openscapes form. In that case use those responses instead.
:::

2. In R, do the following:

  ```r
  library(kyber)
  library(datapasta)
  
  # Run datapasta::tribble_paste() or use the RStudio Addin 'Paste as tribble'
  
  # assign the pasted output to a new object:
  responses <- tribble::tribble(
      # The stuff that was pasted
  )
  ```
  
  The `responses` object will have a column of usernames. It likely won't be 

>   called `username`, so you can rename it orm even easierm change the name inside

Reviewer: ```suggestion
  called `username`, so you can rename it or even easier, change the name inside
```
